From 38437bde77bc141f714b09bffb6b70f29a2c0b75 Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@freescale.com>
Date: Tue, 15 Sep 2015 15:25:12 +0800
Subject: [PATCH] MLK-11552 mmc: sdhci: call sdhci_init() before request irq
 in sd resume

In sdhci_init(), all irqs will be cleared, and then set the needed irqs.
So sdhci_init() should be called before request irq.

If not, some irqs may be triggled wrongly. For example, in i.MX7D LPSR
mode, when system resume, the reset value of register INT_SIGNAL_EN is
not zero, bit 8 is 1, which means SDIO card interrupt enable. And in this
case, the card still has no voltage supply, so the Data3 pin is low,
which always trigger the sd controller rise interrupt. Due to we already
request the irq, so system will always try to do the irq service, and has
no response to other event.

So this patch make sure only set the needed irqs before request irq.

Signed-off-by: Haibo Chen <haibo.chen@freescale.com>
---
 drivers/mmc/host/sdhci.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 09ae0c0..7a87928 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -2696,16 +2696,6 @@ int sdhci_resume_host(struct sdhci_host *host)
 			host->ops->enable_dma(host);
 	}
 
-	if (!device_may_wakeup(mmc_dev(host->mmc))) {
-		ret = request_irq(host->irq, sdhci_irq, IRQF_SHARED,
-				  mmc_hostname(host->mmc), host);
-		if (ret)
-			return ret;
-	} else {
-		sdhci_disable_irq_wakeups(host);
-		disable_irq_wake(host->irq);
-	}
-
 	if ((host->mmc->pm_flags & MMC_PM_KEEP_POWER) &&
 	    (host->quirks2 & SDHCI_QUIRK2_HOST_OFF_CARD_ON)) {
 		/* Card keeps power but host controller does not */
@@ -2718,6 +2708,16 @@ int sdhci_resume_host(struct sdhci_host *host)
 		mmiowb();
 	}
 
+	if (!device_may_wakeup(mmc_dev(host->mmc))) {
+		ret = request_irq(host->irq, sdhci_irq, IRQF_SHARED,
+				  mmc_hostname(host->mmc), host);
+		if (ret)
+			return ret;
+	} else {
+		sdhci_disable_irq_wakeups(host);
+		disable_irq_wake(host->irq);
+	}
+
 	sdhci_enable_card_detection(host);
 
 	if (host->ops->platform_resume)
-- 
1.8.0

